File: app/channels/battle_channel.rb
class BattleChannel < ApplicationCable::Channel
  def subscribed
    stream_from "battle_channel"
  end

  def speak(data)
    ActionCable.server.broadcast "battle_channel", message: data["message"]
  end

  def end_turn(data)
    battle = Battle.find(data["battle_id"])
    board = battle.boards.last

    # 新しいボードを作成（ここは実際のゲームロジックに応じて調整）
    new_board = board.dup
    new_board.turn += 1
    new_board.active_player = board.active_player == 1 ? 2 : 1
    new_board.save!

    # 更新された盤面をブロードキャスト
    html = ApplicationController.renderer.render(
      partial: "battles/board",
      locals: {
        battle: battle,
        board: new_board,
        cards: Card.all.index_by(&:id)
      }
    )

    ActionCable.server.broadcast(
      "battle_channel",
      {
        type: "board_update",
        html: html
      }
    )
  end

  def play_card(data)
    battle = Battle.find(data["battle_id"])
    board = battle.boards.last

    # 新しいボードを作成
    new_board = board.dup

    # プレイヤーと位置に基づいてカードを配置
    player = data["player"].to_i
    position = data["position"]
    sequence = data["sequence"].to_i

    # カードを場に配置
    member_position = "player#{player}_#{position}_member"
    new_board[member_position] = sequence

    # 山札からカードを除去
    deck = player == 1 ? new_board.player1_deck : new_board.player2_deck
    deck.delete(sequence)

    if player == 1
      new_board.player1_deck = deck
    else
      new_board.player2_deck = deck
    end

    new_board.save!

    # 更新された盤面をブロードキャスト
    html = ApplicationController.renderer.render(
      partial: "battles/board",
      locals: {
        battle: battle,
        board: new_board,
        cards: Card.all.index_by(&:id)
      }
    )

    ActionCable.server.broadcast(
      "battle_channel",
      {
        type: "board_update",
        html: html
      }
    )
  end
end

File: app/javascript/channels/battle_channel.js
import consumer from "channels/consumer"

consumer.subscriptions.create("BattleChannel", {
 connected() {
   document.
     querySelector('input[data-behavior="battle_speaker"]').
     addEventListener('keypress', (event) => {
       if (event.key === 'Enter') {
         this.speak(event.target.value);
         event.target.value = '';
         return event.preventDefault();
       }
     });

   // ターン終了ボタンのイベントリスナーを追加
   document.
     querySelector('button[data-behavior="end_turn"]')?.
     addEventListener('click', (event) => {
       const battleId = event.target.dataset.battleId;
       this.endTurn(battleId);
       return event.preventDefault();
     });

   this.setupCardListeners();
 },

 disconnected() {
 },

 received(data) {
   if (data.type === 'board_update') {
     // ボード全体を更新
     const element = document.querySelector('.battle-board');
     element.outerHTML = data.html;
     
     // イベントリスナーを再設定
     this.setupEndTurnButton();
     this.setupCardListeners();
   } else {
     // 既存のメッセージ処理
     const element = document.querySelector('#battle');
     element.insertAdjacentHTML('beforeend', data['message']);
   }
 },

 speak: function(message) {
   return this.perform('speak', { message: message });
 },

 endTurn: function(battleId) {
   return this.perform('end_turn', { battle_id: battleId });
 },

 playCard: function(battleId, sequence, player, position) {
   return this.perform('play_card', {
     battle_id: battleId,
     sequence: sequence,
     player: player,
     position: position
   });
 },

 setupEndTurnButton: function() {
   const button = document.querySelector('button[data-behavior="end_turn"]');
   if (button) {
     button.addEventListener('click', (event) => {
       const battleId = event.target.dataset.battleId;
       this.endTurn(battleId);
       return event.preventDefault();
     });
   }
 },

 setupCardListeners: function() {
   const modal = document.getElementById('position-select-modal');
   let currentCard = null;

   // プレイ可能なカードにクリックイベントを設定
   document.querySelectorAll('[data-behavior="playable_card"]').forEach(card => {
     card.addEventListener('click', (event) => {
       currentCard = event.currentTarget;
       modal.style.display = 'block';
     });
   });

   // モーダル内のボタンにイベントを設定
   document.querySelectorAll('.position-button').forEach(button => {
     button.addEventListener('click', () => {
       if (currentCard) {
         const battleId = document.querySelector('[data-battle-id]').dataset.battleId;
         const sequence = currentCard.dataset.sequence;
         const player = currentCard.dataset.player;
         const position = button.dataset.position;

         this.playCard(battleId, sequence, player, position);
         modal.style.display = 'none';
         currentCard = null;
       }
     });
   });

   // モーダルの外側をクリックした時に閉じる
   modal.addEventListener('click', (event) => {
     if (event.target === modal) {
       modal.style.display = 'none';
       currentCard = null;
     }
   });
 }
});
